DDP-5593: Modal Activity Block

EdPokhvalin opened the pull request:
Final version of the review (without extra changes).

I added a new component with a card and a modal dialog. I'm not sure about the tests, because it's more like a toolkit component with high-level logic, but if we need them, please let me know.
I also have an idea about a second modal window with an arrow. I haven't seen any popovers like this in our projects, so this is the first time. Maybe it would be helpful if I move it to the SDK for reuse?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)
<!-- Reviewable:end -->

EdPokhvalin committed:
DDP-5593 Create modal activity block component

EdPokhvalin committed:
DDP-5593 Add logic and styles to the modal activity dialog

MocanaAtBroad reviewed (changes requested):
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 1 of 3 files reviewed, 7 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, and @fat23cat)

*[a discussion](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFxRvw-ihSx15P2hwM:-MUFxRvx1sVSsdZ2MGwm:b-sbqhfk) (no related file):*
We are missing adding the new component to the rendering of activities. Right now you have created but it sits unused.

---
*[a discussion](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFy0ZMBlWD2oaBIqwj:-MUFy0ZMBlWD2oaBIqwk:b-dku92q) (no related file):*
Also the shape/implementation of the component is quite different than some other "block" component. Please take a look at groupBlock.component.ts , activityQuestion.component.ts, etc and how they are used within activitySection.component.ts

---
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 126 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtud3-k_fvbuDy1WZ:-MUFtud3-k_fvbuDy1W_:b-yoesxa) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L126)):*
> ```XML
>       height: 40px;
>       width: 140px;
>     }`
> ```

I believe we have putting the more detailed styling into the scss configuration and just keeping within the component the more essential CSS that is needed for basic display.
Could you and @alextanlan discuss best way to set the styles? 
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 161 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvLN62kgIuJ6yVoFz:-MUFvLN62kgIuJ6yVoG-:b-6sds36) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L161)):*
> ```XML
>                 <line x1="2.15207317" y1="13.9521951" x2="6.53902439" y2="18.3314634"></line>
>               </g>
>             </svg>
> ```

I believe we have being putting SVGs into separate files rather than embedding them into markup. Will make this more compact and easier to read.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 249 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtBVx7GjExXl-gi7_:-MUFtBVx7GjExXl-gi7a:b-hsob7r) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L249)):*
> ```XML
>       this.config.studyGuid,
>       this.activityGuid
>     ).subscribe(() => this.deleteActivity.emit());
> ```

Should consider adding a pipe(take(1)) to terminate the subscription each time method executed.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 258 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvvJWD1tO59b4rn9G:-MUFvvJWD1tO59b4rn9H:baypwv5) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L258)):*
> ```XML
>       width: `862px`,
>       position: { top: '65px' },
>       panelClass: 'modal-activity-block__edit-dialog',
> ```

Not sure what the best way to handle this, but hard-coding the px width seems problematic when try to execute in mobile.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUFwmmI1e6E0cg2W6Du:b-s0uw2i) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
> ```XML
>     this.activityServiceAgent.getActivitySummary(this.config.studyGuid, this.activityGuid)
>       .subscribe(activityInstance => this.activityInstance = activityInstance);
>   }
> ```

Both of this will need to figure out how to dispose of subscription.


<!-- Sent from Reviewable.io -->

EdPokhvalin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 1 of 3 files reviewed, 7 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 249 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtBVx7GjExXl-gi7_:-MUJ-VAE9VFpmChx-k62:b-1wmezp) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L249)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Should consider adding a pipe(take(1)) to terminate the subscription each time method executed.
</blockquote></details>

I think it is unnecessary in this situation. The *deleteActivityInstance* method uses HttpClient under the hood, and all its observables already auto-unsubscribable. We need to unsubscribe only in case of some custom observables (for example, if we wrap some HTML events in observables or if the observable can run indefinitely).
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUJ2y7YDfmy4koemrFD:b-vcwc8h) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Both of this will need to figure out how to dispose of subscription.
</blockquote></details>

We also use HttpClient here, so I believe we don't need to unsubscribe


<!-- Sent from Reviewable.io -->

alextanlan reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 1 of 3 files reviewed, 12 unresolved discussions (waiting on @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 9 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMsgjr7eOpaifPZs-l:-MUMsgjr7eOpaifPZs-m:b53bkr3) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L9)):*
> ```XML
> 
> @Component({
>   selector: 'modal-activity-block',
> ```

In my believe, keeping this component in ddp-sdk sounds more logical. Like we have other activity blocks components in sdk already. 
Also Marco mentioned in Slack chat: "So given that activity block is part of our API, then we need to have handling of activity block in sdk."
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 11 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMtUdDAnAqc2SYZyMb:-MUMtUdDAnAqc2SYZyMc:be8trn4) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L11)):*
> ```XML
>   selector: 'modal-activity-block',
>   styles: [
>     `.modal-activity-block__card {
> ```

What do you think about saving the styles (and the template as well) in separate files ? 
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 91 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMu3xZ7K-dShvvD5-d:-MUMu3xZ7K-dShvvD5-e:b-qgguph) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L91)):*
> ```XML
>     }`,
>     `::ng-deep .modal-activity-block__delete-dialog > mat-dialog-container {
>       padding: 27.5px 45.5px 20px 45.5px;
> ```

I suppose "a half of a pixel" is overkill.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 126 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtud3-k_fvbuDy1WZ:-MUMuT9a5QWIVxPY4lEh:bcard1k) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L126)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

I believe we have putting the more detailed styling into the scss configuration and just keeping within the component the more essential CSS that is needed for basic display.
Could you and @alextanlan discuss best way to set the styles? 
</blockquote></details>

Marco, I see your point: having a minimum styles in a 'core' reused component and adding all more needed styles in a container (a study page). 
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 161 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvLN62kgIuJ6yVoFz:-MUMvP1w1JR4bkbxkqJa:bmps1dk) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L161)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

I believe we have being putting SVGs into separate files rather than embedding them into markup. Will make this more compact and easier to read.
</blockquote></details>

I guess we already agreed to have svg icons in assets, did we ?
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 211 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMxwAVEeEF5QXZWWiV:-MUMxwAVEeEF5QXZWWiW:b-dvjqz) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L211)):*
> ```XML
>     <ng-template #delete_dialog>
>       <div class="delete-dialog">
>         <span class="delete-dialog__title">Are you sure you want to delete\n this information?</span>
> ```

Should we think about i18n for every text content in the component?
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 258 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvvJWD1tO59b4rn9G:-MUMyjZ159vTfA9k2fwX:bb5qlin) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L258)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Not sure what the best way to handle this, but hard-coding the px width seems problematic when try to execute in mobile.
</blockquote></details>

Perhaps it is worth using 'vh, vw' (for all dialogs) ?
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 269 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMzQs99So6g-O7FgRv:-MUMzQs99So6g-O7FgRw:b-dwl1zc) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L269)):*
> ```XML
>   public openDeleteDialog(): void {
>     const dialogWidth = 396;
>     const realDialogWidth = dialogWidth + 20; // because of the arrow
> ```

What is the arrow here ? (in the comment)
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUMzhPGAMOCHegey6Ti:b-nzh5qq) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

We also use HttpClient here, so I believe we don't need to unsubscribe
</blockquote></details>

I agree with Ed up to a point. Nevertheless, the best practice is using `unsubscribe` (e.g. takeUntil), albeit it can be not needed in that particular case. The request handler can be extended with more logic and it can possibly lead to some side-effects.


<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
DDP-5593 Move modal activity block under the sdk

EdPokhvalin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 0 of 3 files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 9 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMsgjr7eOpaifPZs-l:-MUO6udLCZsISAAeN9xF:bb6un1j) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L9)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

In my believe, keeping this component in ddp-sdk sounds more logical. Like we have other activity blocks components in sdk already. 
Also Marco mentioned in Slack chat: "So given that activity block is part of our API, then we need to have handling of activity block in sdk."
</blockquote></details>

Done
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 269 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMzQs99So6g-O7FgRv:-MUO7g8J33iqoN8V5YsC:bgsi1h4) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L269)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

What is the arrow here ? (in the comment)
</blockquote></details>

This is the arrow at the bottom of the popover that points to its binding to an element on the page. Unfortunately, I was not able to make it so that the width of the arrow is not counted as part of the width of the popup.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUO76pv8_Rjvozl4AkK:b-z44j18) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

I agree with Ed up to a point. Nevertheless, the best practice is using `unsubscribe` (e.g. takeUntil), albeit it can be not needed in that particular case. The request handler can be extended with more logic and it can possibly lead to some side-effects.
</blockquote></details>

I found a good article on where we should and where we shouldn't unsubscribe:
https://netbasal.com/when-to-unsubscribe-in-angular-d61c6b21bad3


<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
DDP-5593 Modal activity block restructuring:
- Move the component to its own directory
- Move the styles and template to separate files
- Add some css variables to customize the appearance of the component
- Use studyGuid from input

alextanlan reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 0 of 5 files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 269 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMzQs99So6g-O7FgRv:-MUOTSx-BPfZ9FfSeKq3:b-arxpra) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L269)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

This is the arrow at the bottom of the popover that points to its binding to an element on the page. Unfortunately, I was not able to make it so that the width of the arrow is not counted as part of the width of the popup.
</blockquote></details>

I see. It just sounded for me as some `magic`world.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUORg4RAjW4-GOWYK0l:bmgx8rw) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

I found a good article on where we should and where we shouldn't unsubscribe:
https://netbasal.com/when-to-unsubscribe-in-angular-d61c6b21bad3
</blockquote></details>

I found not less good discussion:
https://stackoverflow.com/questions/35042929/is-it-necessary-to-unsubscribe-from-observables-created-by-http-methods.
Worth starting reading from "What are you people talking about!!!".
As I said before, I agree with you that in our particular case it's not so necessary. It just seems to me as being a good practice to have some additional conditional as to prevent some `bad` stuff, like e.g. checking an input data for isNull.


<!-- Sent from Reviewable.io -->

EdPokhvalin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 0 of 5 files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 11 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMtUdDAnAqc2SYZyMb:-MUOKIoCEdSXtjhQMcRz:b-886gsx) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L11)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

What do you think about saving the styles (and the template as well) in separate files ? 
</blockquote></details>

Fixed
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 91 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMu3xZ7K-dShvvD5-d:-MUOKKKv5NJDwIi6Lf5Q:b5u6if4) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L91)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

I suppose "a half of a pixel" is overkill.
</blockquote></details>

Also fixed


<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
DDP-5593 Add svg icons to assets

MocanaAtBroad reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 0 of 5 files reviewed, 16 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 1 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUJadH4W_wbK1gjVD_:-MUUJadH4W_wbK1gjVDa:b62qv3) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L1)):*
> ```HTML
> <mat-card class="modal-activity-block__card">
> ```

We have been naming files using camel case. Please rename to make consistent with rest of code.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 15 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUK4e8CYjejLd1yzZF:-MUUK4e8CYjejLd1yzZG:bb7gyar) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L15)):*
> ```HTML
>            class="modal-activity-block__success-message">
>         <mat-icon inline>check_circle</mat-icon>
>         <span>{{activityInstance?.numQuestionsAnswered}} questions answered</span>
> ```

i18n
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 71 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUF7Y3Ez2Z37nL8D6L:-MUUF7Y3Ez2Z37nL8D6M:bopqieb) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L71)):*
> ```HTML
>                           [studyGuid]="studyGuid"
>                           [activityGuid]="activityGuid"
>     ></ddp-activity-section>
> ```

You sure ddp-activity-section is the right component here? I imagine you have not been able to test this out yet? Kind of difficult to test out right now.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 76 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUDhYSEcGjqQOZUf7U:-MUUDhYSEcGjqQOZUf7V:b-hjmdxf) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L76)):*
> ```HTML
>             class="button_primary modal-activity-block__save_activity"
>             (click)="closeEditDialog()">
>       Save
> ```

Let's i18n this and all the other strings in template. Follow pattern used in other components. (userPreferences.component.ts is one example)
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 211 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUMxwAVEeEF5QXZWWiV:-MUUDG0f2-KWPsJN0SB7:b-xhqzki) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L211)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

Should we think about i18n for every text content in the component?
</blockquote></details>

Yes. Everything we do should be i18n-ready.
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUUFiCLC0EqaDBbtnON:b-ep9efe) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

I found not less good discussion:
https://stackoverflow.com/questions/35042929/is-it-necessary-to-unsubscribe-from-observables-created-by-http-methods.
Worth starting reading from "What are you people talking about!!!".
As I said before, I agree with you that in our particular case it's not so necessary. It just seems to me as being a good practice to have some additional conditional as to prevent some `bad` stuff, like e.g. checking an input data for isNull.
</blockquote></details>

@edpokhvalin I believe that you are correct that subscription will be automatically disposed of.
Still, I am inclined to say let's assume we always need to dispose of subscription except for `async` pipes in templates. 
In my mind it is just simpler to assume you need to cleanup than to try to figure out if a particular case is the exception that does not require it.
I look forward to day when rxjs or Angular simply says : cleanup of subscription is not automatic for all subscriptions. In the meantime all subscriptions are suspicious.


<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
DDP-5593:
- Use block as an input
- Support multiple instances under the block
- Use svg icons from assets

EdPokhvalin committed:
DDP-5593 Add modal activity block to the activity section

EdPokhvalin committed:
DDP-5593 Rename a component in the camel case

EdPokhvalin committed:
DDP-5593 Unsubscribe from all observables manually

EdPokhvalin committed:
DDP-5593 Add IDs for i18n

EdPokhvalin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 0 of 8 files reviewed, 16 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

*[a discussion](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFxRvw-ihSx15P2hwM:-MUhqPh15umWs0VAWwTp:b-b1ts4d) (no related file):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

We are missing adding the new component to the rendering of activities. Right now you have created but it sits unused.
</blockquote></details>

I added this component to the activity section, hope I got it right

---
*[a discussion](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFy0ZMBlWD2oaBIqwj:-MUhvtQ29Jxj3LpvAPlW:b-ktagle) (no related file):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Also the shape/implementation of the component is quite different than some other "block" component. Please take a look at groupBlock.component.ts , activityQuestion.component.ts, etc and how they are used within activitySection.component.ts
</blockquote></details>

I think I understand you, Marco. Now the component will work with the block instead of the instance. Is that what you meant?

---
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 1 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUJadH4W_wbK1gjVD_:-MUhtfmzCLdwdEgnQ_2C:b-886gsx) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L1)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

We have been naming files using camel case. Please rename to make consistent with rest of code.
</blockquote></details>

Fixed
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 15 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUK4e8CYjejLd1yzZF:-MUhyqhU1x4JXGyII0k9:b-w1wb03) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L15)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

i18n
</blockquote></details>

All i18n-related stuff is done or at least I think so)
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 76 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUDhYSEcGjqQOZUf7U:-MUhthPgBV2V4QsfIaDw:b-bmqxzj) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L76)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Let's i18n this and all the other strings in template. Follow pattern used in other components. (userPreferences.component.ts is one example)
</blockquote></details>

Yeah, I've already looked at it and I have a couple of questions:
* Do I need to add all the default translations in the configurations in all our apps? I just added a IDs to those translations
* I'm not sure about label like "# questions answered". In different languages, the number can stand in different places. Should I use a parameter like here?
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 126 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtud3-k_fvbuDy1WZ:-MUOi7oeBxGdJpCNUqcV:boxo8gt) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L126)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

Marco, I see your point: having a minimum styles in a 'core' reused component and adding all more needed styles in a container (a study page). 
</blockquote></details>

I also added some CSS variables to simplify the customization
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 161 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvLN62kgIuJ6yVoFz:-MUhqXTb3uV-INA6efTR:b-bw57du) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L161)):*
<details><summary><i>Previously, alextanlan wrote…</i></summary><blockquote>

I guess we already agreed to have svg icons in assets, did we ?
</blockquote></details>

SVGs are now taken from assets
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 249 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFtBVx7GjExXl-gi7_:-MUhyc5y7wb1ka_6rIZd:blf7sia) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L249)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

I think it is unnecessary in this situation. The *deleteActivityInstance* method uses HttpClient under the hood, and all its observables already auto-unsubscribable. We need to unsubscribe only in case of some custom observables (for example, if we wrap some HTML events in observables or if the observable can run indefinitely).
</blockquote></details>

Fixed with *take(1)*
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 301 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFwmmI1e6E0cg2W6Dt:-MUhrtZK1aIEdL4ClLSN:b8cfioo) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L301)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

@edpokhvalin I believe that you are correct that subscription will be automatically disposed of.
Still, I am inclined to say let's assume we always need to dispose of subscription except for `async` pipes in templates. 
In my mind it is just simpler to assume you need to cleanup than to try to figure out if a particular case is the exception that does not require it.
I look forward to day when rxjs or Angular simply says : cleanup of subscription is not automatic for all subscriptions. In the meantime all subscriptions are suspicious.
</blockquote></details>

Ok, I got your point, colleagues. Now all subscriptions will be closed manually


<!-- Sent from Reviewable.io -->

MocanaAtBroad reviewed (dismissed):
> Reviewed 7 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 7 of 8 files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 76 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUDhYSEcGjqQOZUf7U:-MUjbFRMB1gRfj79jBYx:b-vmbvi5) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L76)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

Yeah, I've already looked at it and I have a couple of questions:
* Do I need to add all the default translations in the configurations in all our apps? I just added a IDs to those translations
* I'm not sure about label like "# questions answered". In different languages, the number can stand in different places. Should I use a parameter like here?
</blockquote></details>

Yes. Using a parameter is a good idea.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modalActivityBlock/modalActivityBlock.component.ts, line 2 at r4](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUjeCp57gyn9QbCI7U5:-MUjeCp57gyn9QbCI7U6:b-s9vfx4) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/17c433e77dc400a84edecee4382d3297d3345975/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modalActivityBlock/modalActivityBlock.component.ts#L2)):*
> ```XML
> import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, OnInit, Output, TemplateRef, ViewChild } from '@angular/core';
> import { ActivityForm, ActivityInstance, ActivityServiceAgent } from 'ddp-sdk';
> ```

I think this line might be causing a circular dependency warning when I try to `ng serve`. Importing the ddp-sdk from a component in ddp-sdk is problematic
Please change imports of these classes to relative paths.


<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
DDP-5593 Use a full import address

EdPokhvalin committed:
DDP-5593 Fix message

EdPokhvalin committed:
Merge branch 'develop' of https://github.com/broadinstitute/ddp-angular into DDP-5593-modal-activity-blocks

EdPokhvalin committed:
DDP-5593 Use child activity guid

EdPokhvalin committed:
DDP-5593 Use relative units

EdPokhvalin committed:
DDP-5593 Add labels to osteo

EdPokhvalin committed:
DDP-5593 Add a loading spinner

EdPokhvalin committed:
DDP-5593:
- update the instance after closing the dialog box
- simplify the appearance of the delete dialog box when working on a mobile device

EdPokhvalin committed:
DDP-5593 Delete unnecessary import

EdPokhvalin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: 4 of 9 files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html, line 71 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUUF7Y3Ez2Z37nL8D6L:-MVSDUO_3xUhBtgdMMME:b6nv79f) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/f9e2a5b8dd5d55acceb925fb44b955d928975fde/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modal-activity-block/modal-activity-block.component.html#L71)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

You sure ddp-activity-section is the right component here? I imagine you have not been able to test this out yet? Kind of difficult to test out right now.
</blockquote></details>

It's working now
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modalActivityBlock/modalActivityBlock.component.ts, line 2 at r4](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUjeCp57gyn9QbCI7U5:-MVRWe535nI3bpjEE33C:bb6un1j) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/17c433e77dc400a84edecee4382d3297d3345975/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/modalActivityBlock/modalActivityBlock.component.ts#L2)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

I think this line might be causing a circular dependency warning when I try to `ng serve`. Importing the ddp-sdk from a component in ddp-sdk is problematic
Please change imports of these classes to relative paths.
</blockquote></details>

Done
___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 258 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvvJWD1tO59b4rn9G:-MVSDQh4DGckz7Bl2Zh9:b-886gsx) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L258)):*
<details><summary><i>Previously, alextanlan (Alex Maratkanov) wrote…</i></summary><blockquote>

Perhaps it is worth using 'vh, vw' (for all dialogs) ?
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

alextanlan reviewed (approved):
> Reviewed 1 of 6 files at r3, 6 of 9 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: all files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)



<!-- Sent from Reviewable.io -->

EdPokhvalin committed:
Merge branch 'develop' of https://github.com/broadinstitute/ddp-angular into DDP-5593-modal-activity-blocks

alextanlan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/456)* status: all files reviewed, 12 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts, line 258 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/456#-MUFvvJWD1tO59b4rn9G:-MVWilHh5pGe7Yk2tSzv:btojbfq) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/863086b3cc387b7d015140281f03931462cf2c61/ddp-workspace/projects/toolkit/src/lib/components/activity/modal-activity-block.component.ts#L258)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

Fixed
</blockquote></details>

Great.


<!-- Sent from Reviewable.io -->
